### ISSUE TITLE
gpt3_mcore_te_tp2_zp_z3_resume_torch_dist_te_8experts2parallel_top2router broken after TE2.14

### ISSUE THREAD

**Issue by ko3n1g** [NVIDIA/Megatron-LM#4033]
**Describe the bug**

Captured via CI run: https://github.com/NVIDIA/Megatron-LM/actions/runs/23531562829/job/68506373685?pr=4025

I would assume we need to update the checkpoint, but that's just a hunch I'm having.

**Steps/Code to reproduce bug**

Please list *minimal* steps or code snippet for us to be able to reproduce the bug.

A helpful guide on on how to craft a minimal bug report http://matthewrocklin.com/blog/work/2018/02/28/minimal-bug-reports.

**Expected behavior**

**Additional context**

**Comment by DhineshPonnarasan** [comment]
## Root Cause
Checkpoints created with mcore ≤ 0.14 include an optimizer state attribute (`flattened_range`) that is no longer supported in newer versions. Loading these checkpoints will fail for the optimizer state, but model weights remain compatible.

## Proposed Solution

### Workaround
- When resuming from an old checkpoint, use the `--no-load-optim` flag to skip loading the optimizer state.
- This allows training to resume from the checkpointed model weights with a freshly initialized optimizer.

### Code Improvement
- Add error handling in the code to detect and warn about this incompatibility, suggesting the use of `--no-load-optim`.

### Documentation
- Update documentation to clarify that optimizer states from mcore ≤ 0.14 are not supported.
- Recommend the workaround.

### (Optional)
- Provide a migration script to convert or strip old optimizer states if legacy support is required.

## Example Patch
```python
# In megatron/training/checkpointing.py, after loading the optimizer state:
if optimizer is not None and 'flattened_range' in optimizer_state_dict:
    raise RuntimeError(
        "Optimizer state from checkpoints created with mcore <= 0.14 is not supported. "
        "Please resume with --no-load-optim to skip loading the optimizer state."
    )

---

Hi @ko3n1g ,
I would like to work on implementing this fix.

**Comment by ko3n1g** [comment]
## Investigation update (2026-05-26)

Reproduced the current state of the test and reviewed the change history.

**Test:** `tests/functional_tests/test_cases/moe/gpt3_mcore_te_tp2_zp_z3_resume_torch_dist_te_8experts2parallel_top2router/`

**Current status:** Disabled in both H100 and GB200 recipes:
- `tests/test_utils/recipes/h100/moe.yaml:171-175` — `scope: [mr-broken, mr-github-broken]`
- `tests/test_utils/recipes/gb200/moe.yaml:150-154` — `scope: [nightly-broken]`

It has been in this state since the file was first introduced on 2026-05-01 (commit `3b1521e29c`).

**Re: the earlier root-cause comment** (`flattened_range` in old checkpoints): I don't think that applies here. The test is `TEST_TYPE: ckpt-resume`, so `tests/functional_tests/shell_test_utils/run_ci_test.sh:248` forces `CHECKPOINT_LOAD_PATH=/tmp/checkpoints/` for the first run — there is no external pre-saved checkpoint. Both training runs save and load checkpoints produced by the same MCore version that ran the test. `--no-load-optim` would not be a correct fix.

**Likely actual cause:** Numerical drift from the TE 2.14 bump (`ac6ca5b299` / `ceac269466` / `b20c1b25e5`). The same pattern was handled for sibling tests by simply refreshing golden values, e.g. `8821e6f633` ("fix(ci): re-enable scoped_cudagraph MoE test with TE 2.14 golden values", #4444). The golden values in this test case predate the TE 2.14 bump and have not been touched since.

**Why PR #4886 doesn't already cover this:** The container bump / golden refresh PR (#4886) refreshes goldens for 60+ functional tests including the closely-related `gpt3_mcore_te_tp2_zp_z3_resume_fsdp_dtensor*`, but it skipped this one specifically — presumably because the test was already `-broken` in CI before the refresh run.

**Proposed fix path (requires H100 CI access):**
1. Flip both recipes from `mr-broken/mr-github-broken` -> `mr/mr-github` (H100) and `nightly-broken` -> `nightly` (GB200).
2. Trigger a functional-test run with the "Run functional tests" label.
3. Refresh `golden_values_dev_dgx_h100.json` and `golden_values_dev_dgx_gb200.json` using `tests/test_utils/python_scripts/download_golden_values.py --source github --pipeline-id <run-id>`.
4. If the test still fails on goldens, bisect against the TE bump commits to confirm it is pure numerical drift and not a behavioral regression.

I am unable to complete steps 2-3 from this environment (no H100/GB200 CI bound to my fork, and the hard rules prohibit committing speculative goldens). Leaving the issue open and assigned to @janEbert pending a CI-equipped run.
